keps/sig-api-machinery/4933-cacher-label-index/README.md

```diff
@@ -0,0 +1,875 @@
+
+### Commandline changes
+
+Considering that label names may be different in different scenarios, such as app label name in Flink and spark-app-selector label name in Spark, this feature will add a commandline parameter to the apiserver, allowing different label names to be configured as the trigger index of the apiserver cacher.
+
+```
+--index-labels. Index labels settings for some resources (pods, nodes, etc.), comma separated. The individual setting format: resource[.group]#labelName1;labelName2, where resource is lowercase plural (no version), group is omitted for resources of apiVersion v1 (the legacy core API) and included for others, and labelName is a string. It takes effect when watch-cache and feature gate CacherLabelIndex are enabled. In addition to built-in indexes such as pod.spec.nodeName, apiserver cacher will create additional indexes for these labels to speed up list and watch requests.
+
+# e.g. for spark
+--index-labels=pods#spark-app-selector
+# e.g. for spark and flink
+--index-labels=pods#spark-app-selector;app
+```
+
+### Changes of register trigger index and cache index to resource storage
+
+In order to passing the commandline label indexes parameters into resource storage and cacher, we'll add a `IndexLabels []string` field into `Config` struct in file staging/src/k8s.io/apiserver/pkg/storage/storagebackend/config.go
+
+We will parse the commandline label indexes parameters in `Store.CompleteWithOptions` and register them into `Indexers` and `TriggerFunc` of `StoreOptions`.
+
+### Major struct and field changes in cacher
+
+Currently `indexedTrigger *indexedTriggerFunc` is a single function field in cacher to support field index. It is assigned to the `spec.nodeName` function for pod resource.
+
+We will update it to `indexedTriggers map[string]storage.IndexerFunc` to support a single field index and multiple label indexes. Multiple field indexes is not supported as before and is beyond the scope of this KEP.
+
+Then we will add a `triggerIndex` struct and change `valueWatchers map[string]watchersMap` of `indexedWatchers` struct to `valueWatchers map[triggerIndex]watchersMap` so we can add `app`, `k8s-app`, `spark-app-selector` or any other label names as trigger functions. 
+
+```golang
+type triggerIndex struct {
+	indexName string
+	value     string
+}
+```
+
+### Changes of event processing flow in cacher
+
+In the process of filtering cacheWatchers in phase 2, for a pod change event, currently only the unique field triggerIndex is judged. We will modify it to try to match all registered field and label triggerIndexes. And then send the pod event to the cache watchers corresponding to the value watchers matching these triggerIndex.
```

Comment:
I don't understand this - can you please clarify?

Reply:
This describes all the code changes of https://github.com/kubernetes/kubernetes/pull/126284/files#diff-14c4fb9f290753f50e0af4856d871ff41e2b520d747158cbff7bafc577fbb29eR590-R615
and
https://github.com/kubernetes/kubernetes/pull/126284/files#diff-14c4fb9f290753f50e0af4856d871ff41e2b520d747158cbff7bafc577fbb29eR1032-R1047